x-pack/plugin/esql-core/src/main/java/org/elasticsearch/xpack/esql/core/tree/Source.java

```diff
@@ -25,6 +33,31 @@ public Source(Location location, String text) {
         this.text = text;
     }
 
+    public static Source readFrom(StreamInput in) throws IOException {
```

Comment:
I think the method signature should state more clearly that this doesn't work with a general `StreamInput`.
```suggestion
    public static <S extends StreamInput & PlanStreamInput> Source readFrom(S in) throws IOException {
```
Then we also don't need to risk the `ClassCastException` below.

Reply:
PlanStreamInput is a StreamInput. So the type could be a PlanStreamInput, no? Or does that force the caller to do the cast ?

Reply:
I'll give this a go. My instincts are that things like this in java don't work well.

Reply:
There's an abstract superclass vs interface thing in the way. It's hopefully temporary, if confusing.